internal/controller/kubecertagent/kubecertagent.go

```diff
@@ -268,42 +274,51 @@ func (c *agentController) Sync(ctx controllerlib.Context) error {
 		return c.failStrategyAndErr(ctx.Context, credIssuer, err, configv1alpha1.CouldNotFetchKeyStrategyReason)
 	}
 
-	if err := c.createOrUpdateDeployment(ctx, newestControllerManager); err != nil {
-		err := fmt.Errorf("could not ensure agent deployment: %w", err)
-		return c.failStrategyAndErr(ctx.Context, credIssuer, err, configv1alpha1.CouldNotFetchKeyStrategyReason)
+	depErr := c.createOrUpdateDeployment(ctx, newestControllerManager)
+	if depErr != nil {
+		// it is fine if this call fails because a different concierge pod may have already created a compatible deployment
+		// thus if the later code is able to find pods with the agent labels that we expect, we will attempt to use them
+		// this means that we must always change the agent labels when we change the agent pods in an incompatible way
+		depErr = fmt.Errorf("could not ensure agent deployment: %w", depErr)
```

Comment:
@enj I'm not sure if I understand this change. If during an upgrade the new leader controller is unable to update the agent pod deployment, then that could indicate a serious problem/bug in the code. Won't this paper over it by allowing the old deployment that was created by the old version of Pinniped to carry on being used indefinitely, despite the changes not actually being applied? It seems like with this change we would need to update the agent pod label anytime we update anything about how this controller defines the agent deployment... not only for breaking changes but basically for all changes. Maybe that was the intention?

Reply:
> It seems like with this change we would need to update the agent pod label anytime we update anything about how this controller defines the agent deployment... not only for breaking changes but basically for all changes.

Any change in the new pinniped that changes the meaning of the kube cert agent as seen by the exec logic of an old pinniped is a breaking change.  So if we decide to add some annotations or similar (say we change the readiness checks) to the deployment, then we do not need to bump the agent labels because we have not changed the semantics of how pods get the signer key (in the readiness example we just changed how the pods are health checked but nothing related to the signer itself).  But like if we change the format of the JSON that the pod emits on the exec call, then we do need to bump the agent label because an old pod may misunderstand the meaning of the JSON and do something unexpected.

> If during an upgrade the new leader controller is unable to update the agent pod deployment, then that could indicate a serious problem/bug in the code. Won't this paper over it by allowing the old deployment that was created by the old version of Pinniped to carry on being used indefinitely, despite the changes not actually being applied?

If the agent label has not changed, the new pinniped code is basically saying that it is safe to use the old deployment forever.  Theoretically it could paper over things like "we broke the RBAC of the new pinniped so it cannot update the deployment" but hopefully we would catch those style of breakages in our CI.  Note that the status stays unhealthy until the new pinniped is able to both update the deployment and fetch the key using it.  Thus it would be clear from the outside that the process is unhappy about its current state.

Reply:
> So if we decide to add some annotations or similar

Yes, that's a good example of what I was concerned about. If a new version of Pinniped wants to add/remove something like an annotation or label or pod setting, but updating the deployment fails for some unexpected reason, then with these changes it would keep using the old deployment which lacks the new annotation/label/setting, and the error is swept under the rug, right?

> If the agent label has not changed, the new pinniped code is basically saying that it is safe to use the old deployment forever.

Yes, that's how I understood this change too. So if the new version of Pinniped actually cares about rolling out the added/removed annotation/label/setting then it must update the label version number, right? If that's the case, then maybe the comment for `agentPodLabelKey` should also describe that subtlety.

Reply:
> the error is swept under the rug, right?

The status remains in an error state meaning things like `pinniped get kubeconfig` would fail.

> describe that subtlety

Updated the comment.  See what you think.

Reply:
Cool, thanks.